cmd/libs/go2idl/go-to-protobuf/protobuf/package.go

```diff
@@ -127,6 +138,43 @@ func (p *protobufPackage) HasGoType(name string) bool {
 	return ok
 }
 
+func (p *protobufPackage) ExtractGeneratedType(t *ast.TypeSpec) bool {
+	if !p.HasGoType(t.Name.Name) {
+		return false
+	}
+
+	switch s := t.Type.(type) {
+	case *ast.StructType:
+		for i, f := range s.Fields.List {
+			if len(f.Tag.Value) == 0 {
+				continue
+			}
+			tag := strings.Trim(f.Tag.Value, "`")
+			protobufTag := reflect.StructTag(tag).Get("protobuf")
+			if len(protobufTag) == 0 {
+				continue
+			}
+			if len(f.Names) > 1 {
+				log.Printf("WARNING: struct %s field %d %s: defined multiple names but single protobuf tag", t.Name.Name, i, f.Names[0].Name)
```

Comment:
I'm OK if we outlaw that...

On Mon, Jan 25, 2016 at 6:46 PM, Clayton Coleman <notifications@github.com>
wrote:

> In cmd/libs/go2idl/go-to-protobuf/protobuf/package.go
> <https://github.com/kubernetes/kubernetes/pull/19426#discussion_r50790202>
> :
>
> > +		return false
> > +	}
> > +
> > +	switch s := t.Type.(type) {
> > +	case *ast.StructType:
> > +		for i, f := range s.Fields.List {
> > +			if len(f.Tag.Value) == 0 {
> > +				continue
> > +			}
> > +			tag := strings.Trim(f.Tag.Value, "`")
> > +			protobufTag := reflect.StructTag(tag).Get("protobuf")
> > +			if len(protobufTag) == 0 {
> > +				continue
> > +			}
> > +			if len(f.Names) > 1 {
> > +				log.Printf("WARNING: struct %s field %d %s: defined multiple names but single protobuf tag", t.Name.Name, i, f.Names[0].Name)
>
> Someone did "Foo, Bar string `protobuf:"baz"`" which means we have the
> same field defined twice.
> … <#-1348009057_>
> On Mon, Jan 25, 2016 at 11:13 AM, Paul Morie <notifications@github.com>
> wrote: In cmd/libs/go2idl/go-to-protobuf/protobuf/package.go <#19426
> (comment)
> <https://github.com/kubernetes/kubernetes/pull/19426#discussion_r50714702>>
> : > + return false > + } > + > + switch s := t.Type.(type) { > + case
> *ast.StructType: > + for i, f := range s.Fields.List { > + if
> len(f.Tag.Value) == 0 { > + continue > + } > + tag :=
> strings.Trim(f.Tag.Value, "`") > + protobufTag :=
> reflect.StructTag(tag).Get("protobuf") > + if len(protobufTag) == 0 { > +
> continue > + } > + if len(f.Names) > 1 { > + log.Printf("WARNING: struct %s
> field %d %s: defined multiple names but single protobuf tag", t.Name.Name,
> i, f.Names[0].Name) Is this supposed to cover the case where someone has a
> field tagged with json:"foo" yaml:"bar" ? — Reply to this email directly or
> view it on GitHub <
> https://github.com/kubernetes/kubernetes/pull/19426/files#r50714702>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/19426/files#r50790202>.
>